File: 2023/16.py
from collections import deque

data = open('data/16.txt').read().splitlines()


def fire_beam(r, c, dr, dc):
    beam = [(r, c, dr, dc)]
    seen = set()
    queue = deque(beam)

    def add(val):
        if val not in seen:
            seen.add(val)
            queue.append(val)

    while queue:
        r, c, dr, dc = queue.popleft()
        r += dr
        c += dc
        if r < 0 or r >= len(data) or c < 0 or c >= len(data[0]):
            continue
        ch = data[r][c]
        if ch == '.' or (ch == '-' and dc != 0) or (ch == '|' and dr != 0):
            add((r, c, dr, dc))
        elif ch == '/':
            dr, dc = -dc, -dr
            add((r, c, dr, dc))
        elif ch == '\\':
            dr, dc = dc, dr
            add((r, c, dr, dc))
        else:
            for dr, dc in [(1, 0), (-1, 0)] if ch == '|' else [(0, 1), (0, -1)]:
                add((r, c, dr, dc))
    return len({(r, c) for (r, c, _, _) in seen})


def solve_part_one():
    return fire_beam(0, -1, 0, 1)


def solve_part_two():
    max_val = 0
    for r in range(len(data)):
        max_val = max(max_val, fire_beam(r, -1, 0, 1))
        max_val = max(max_val, fire_beam(r, len(data[0]), 0, -1))
    for c in range(len(data)):
        max_val = max(max_val, fire_beam(-1, c, 1, 0))
        max_val = max(max_val, fire_beam(len(data), c, -1, 0))
    return max_val


print('Part One: %d' % solve_part_one())
print('Part Two: %d' % solve_part_two())

File: 2017/16.fs
namespace _2017

open System
open System.Collections.Generic

module _16 =

    let Data = (Utils.GetInputData 16).Split(',', StringSplitOptions.RemoveEmptyEntries) |> List.ofArray

    let spin size programs =
        let splitIndex = List.length programs - size
        List.skip splitIndex programs @ List.take splitIndex programs

    let exchange posA posB programs =
        let swap lst i j =
            lst |> List.mapi (fun idx el ->
                if idx = i then lst[j]
                elif idx = j then lst[i]
                else el)
        swap programs posA posB

    let partner charA charB programs =
        let posA = List.findIndex(fun c -> c = charA) programs
        let posB = List.findIndex(fun c -> c = charB) programs
        exchange posA posB programs

    let performMove programs (move: string) =
        match move[0] with
        | 's' ->
            let size = move[1..] |> int
            spin size programs
        | 'x' ->
            let positions = move[1..].Split('/')
            let posA = positions[0] |> int
            let posB = positions[1] |> int
            exchange posA posB programs
        | 'p' ->
            let chars = move[1..].Split('/')
            let charA = chars[0].[0]
            let charB = chars[1].[0]
            partner charA charB programs
        | _ -> programs

    let dance currentState moves = moves |> List.fold performMove currentState

    let rec findCycle currentPrograms input (states: Dictionary<int, char list>) i times =
        if i >= times then currentPrograms
        elif states.ContainsValue(currentPrograms) then states[times % i]
        else
            states[i] <- currentPrograms
            let newPrograms = dance currentPrograms input
            findCycle newPrograms input states (i + 1) times

    let solve (input: string list) (times: int) : string =
        let states = Dictionary<int, char list>()
        let finalState = findCycle ['a'..'p'] input states 0 times
        finalState |> List.map string |> String.concat ""

    let solvePartOne() = solve Data 1

    let solvePartTwo() = solve Data 1000000000

    let Run() =
        printfn $"Part One: {solvePartOne()}"
        printfn $"Part Two: {solvePartTwo()}"